CONTRIBUTING.md

```diff
@@ -1,133 +1,438 @@
-    what Host was used for build (OS, CPU, compiler, ..), what Target was
-    used for verification (arch, board:config, ..), etc.
-    Providing build and runtime logs from before and after change is highly
-    appreciated.
+Description can be single or several sentences long or bullet points
+but enough for anyone to understand change goals and details.
+Usually it will look similar for PR and git commit message.
 
-* Introduce only one functional change per pull-request.
+Good description when read once will make others understand your work.
+This allows faster review, quick understanding of change history (what
+changed, why, how), and eventual problem identification / fix in future.
 
-* Provide finished and verified solutions.
-  Clearly mark Work-In-Progress pull requests not yet ready for assessment.
 
+### 1.4. COMMIT and PR descriptions are equally important.
 
-## References
+Git commit messages are as important as PR descriptions.
 
-### Example Pull Request Report
+Git log provides offline descriptions of each change that is
+git client / interface independent.
 
-#### Summary
+
+### 1.5. COMMIT requirements.
+
+Proper git commit message according to template (see 2.1) is **mandatory**,
+or change is auto-rejected until fixed. Build and runtime logs are
+optional here if these are too long and already provided in the PR.
+
+Git commit message consists of:
+
+ 1. Topic with functional area prefix, `:` mark, short self-explanatory
+    functional context, `.` mark.
+ 2. Blank line.
+ 3. Description on what is changed, how, and why. May use several
+    lines, short sentences, or bullet points (see 1.3).
+ 4. Blank line.
+ 5. Signature (created with `git commit -s`).
+
+Valid git commit example:
+
+```
+net/can: Add g_ prefix to can_dlc_to_len and len_to_can_dlc.
+
+Add g_ prefix to can_dlc_to_len and len_to_can_dlc to
+follow NuttX coding style conventions for global symbols,
+improving code readability and maintainability.
+* you can also use bullet points.
+* to note different thing briefly.
+
+Signed-off-by: AuthorName <Valid@EmailAddress>
+```
+
+
+### 1.6. COMMIT message mandatory fields.
+
+Each git commit message **must** contain these fields, or change is
+auto-rejected until fixed:
+
+ 1. topic.
+ 2. description.
+ 3. signature (`git commit -s`).
+
+Although this seems to repeat rule 1.5, it clearly filters out commits
+with no topic, description, or signature.
+
+
+### 1.7. PR requirements.
+
+1. Proper description (see 1.3) of PR according to template (see 2.3)
+   is **mandatory**. All fields are required or change is auto-rejected
+   until fixed.
+2. For code changes build and runtime logs are **mandatory** to prove code
+   was tested on **at least one** real world hardware target.
+3. Pull Requests should be as small as possible and focused on only one
+   functional change.
+4. Different functional changes must be provided in a separate Pull Requests.
```

Comment:
```suggestion
4. Different logical functions changes must be provided in a separate Pull Requests.

Reply:
how about "Changes affecting different functionalities must be provided in a separate Pull Requests." ? :-)